types/context_cache.go

```diff
@@ -37,73 +31,51 @@ func (c *ContextMemCache) UpsertDeferredSends(moduleAccount string, amount Coins
 	c.deferredBankOpsLock.Lock()
 	defer c.deferredBankOpsLock.Unlock()
 
-	// If there's already a pending withdrawal then subtract it from that amount first
-	// or else add it to the deferredSends mapping
-	ok := c.deferredWithdrawals.SafeSub(moduleAccount, amount)
-	if !ok {
-		c.deferredSends.UpsertMapping(moduleAccount, amount)
-	}
+	c.deferredSends.UpsertMapping(moduleAccount, amount)
 	return nil
 }
 
-func (c *ContextMemCache) SafeSubDeferredSends(moduleAccount string, amount Coins) bool {
+// This will perform a saturating sub against the deferred module balance atomically. This means that it will subtract any balances that it is able to, and then call the passed in `remainder handler` to ensure the remaining balance can be safely subtracted as well. If this remainderHandler returns an error, we will revert the saturating sub to ensure atomicity of the subtraction across the multiple balances
+func (c *ContextMemCache) SaturatingSubDeferredSends(moduleAccount string, amount Coins, remainderHandler func(amount Coins) error) error {
 	if !amount.IsValid() {
 		panic(sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String()))
 	}
-
-	return c.deferredSends.SafeSub(moduleAccount, amount)
-}
-
-func (c *ContextMemCache) RangeOnDeferredSendsAndDelete(apply func(recipient string, amount Coins)) {
-	c.deferredSends.RangeOnMapping(apply)
-}
-
-func (c *ContextMemCache) UpsertDeferredWithdrawals(moduleAccount string, amount Coins) error {
-	if !amount.IsValid() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String())
-	}
-
-	// Separate locks needed for all mappings - atmoic transaction needed
-	c.deferredBankOpsLock.Lock()
-	defer c.deferredBankOpsLock.Unlock()
-
-	// If there's already a pending deposit then subtract it from that amount first
-	// or else add it to the deferredWithdrawals mapping
-	ok := c.deferredSends.SafeSub(moduleAccount, amount)
-	if !ok {
-		c.deferredWithdrawals.UpsertMapping(moduleAccount, amount)
+	originalBalance, ok := c.deferredSends.Get(moduleAccount)
```

Comment:
I think we need to lock/unlock here or else we'll have race conditions:
```
	c.deferredBankOpsLock.Lock()
	defer c.deferredBankOpsLock.Unlock()
```

Reply:
good catch

Reply:
shoudl we consider running go-race with this refactor?

Reply:
we can do it when we start testing this version